src/imap_stream.rs

```diff
@@ -123,160 +77,350 @@ impl<R: Read + Write + Unpin> ImapStream<R> {
-                    return Poll::Ready(Some(Ok(response)));
-                }
-                DecodeResult::None(buffer) => buffer,
-            }
+    /// Returns the number of bytes in the buffer containing data.
+    fn used(&self) -> usize {
+        self.offset
+    }
+
+    /// Returns the unused part of the buffer to which new data can be written.
+    fn free_as_mut_slice(&mut self) -> &mut [u8] {
+        &mut self.block[self.offset..]
+    }
+
+    /// Indicate how many new bytes were written into the buffer.
+    ///
+    /// When new bytes are written into the slice returned by [tail_as_slice] this method
+    /// should be called to extend the used portion of the buffer to include the new data.
+    ///
+    /// You can not write past the end of the buffer, so extending more then there is free
+    /// space marks the entire buffer as used.
+    // aka advance()?
+    fn extend_used(&mut self, num_bytes: usize) {
+        self.offset += num_bytes;
+        if self.offset > self.block.size() {
+            self.offset = self.block.size();
+        }
+    }
+
+    /// Ensure the buffer has free capacity, optionally ensuring minimum buffer size.
+    fn ensure_capacity(&mut self, required: Option<usize>) -> io::Result<()> {
+        let free_bytes: usize = self.block.size() - self.offset;
+        let min_required_bytes: usize = required.unwrap_or(0);
+        let extra_bytes_needed: usize = min_required_bytes.saturating_sub(self.block.size());
+        if free_bytes == 0 || extra_bytes_needed > 0 {
+            let increase = std::cmp::max(Buffer::BLOCK_SIZE, extra_bytes_needed);
+            self.grow(increase)?;
+        }
+        Ok(())
+    }
+
+    /// Grows the buffer, ensuring there are free bytes in the tail.
+    ///
+    /// The specified number of bytes is only a minimum.  The buffer could grow by more as
+    /// it will always grow in multiples of [BLOCK_SIZE].
+    ///
+    /// If the size would be larger than [MAX_CAPACITY] an error is returned.
+    // TODO: This bypasses the byte-pool block re-use.  That's bad.
+    fn grow(&mut self, num_bytes: usize) -> io::Result<()> {
+        let min_size = self.block.size() + num_bytes;
+        let new_size = match min_size % Self::BLOCK_SIZE {
+            0 => min_size,
+            n => min_size + (Self::BLOCK_SIZE - n),
+        };
+        if new_size > Self::MAX_CAPACITY {
+            Err(io::Error::new(
+                io::ErrorKind::Other,
+                "incoming data too large",
+            ))
         } else {
-            buffer
+            self.block.realloc(new_size);
+            Ok(())
+        }
+    }
+
+    /// Return the block backing the buffer.
+    ///
+    /// Next you *must* either return this block using [return_block] or call
+    /// [reset_with_data].
+    // TODO: Enforce this with typestate.
+    fn take_block(&mut self) -> Block<'static> {
+        std::mem::replace(&mut self.block, POOL.alloc(Self::BLOCK_SIZE))
```

Comment:
`replace` does not call `drop`, so the block is not properly returned to the pool (this was likely an issue before as well. So you should do sth like this

```rust
let _ = mem::replace(..);
```
as `replace` returns the value, and the `let _ = ` will trigger the `drop` being evaluated.